operators/src/source/ogr_source/dataset_iterator.rs

```diff
@@ -119,16 +119,69 @@ impl OgrDatasetIterator {
         query_rectangle: &VectorQueryRectangle,
         attribute_filters: &[AttributeFilter],
     ) -> Result<FeaturesProvider<'d>> {
-        // TODO: add OGR time filter if forced
+        let filter_string = if dataset.driver().short_name() == "CSV" {
+            FeaturesProvider::create_attribute_filter_string_cast(attribute_filters)
+        } else {
+            FeaturesProvider::create_attribute_filter_string(attribute_filters)
+        };
+
+        let time_filter = if dataset_information.force_ogr_time_filter {
+            debug!(
+                "using time filter {:?} for layer {:?}",
+                query_rectangle.time_interval, &dataset_information.layer_name
+            );
+            FeaturesProvider::create_time_filter_string(
+                dataset_information.time.clone(),
+                query_rectangle.time_interval,
+                &dataset.driver().short_name(),
+            )
+        } else {
+            None
+        };
+
+        let final_filter = filter_string
+            .map(|f| match &dataset_information.attribute_query {
+                Some(a) => format!("({a}) AND {f}"),
+                None => f,
+            })
+            .or_else(|| dataset_information.attribute_query.clone())
+            .map(|f| match &time_filter {
+                None => f,
+                Some(t) => format!("({t}) AND {f}"),
+            })
+            .or_else(|| time_filter.clone());
 
         let mut features_provider = if let Some(sql) = dataset_information.sql_query.as_ref() {
+            let query = if let Some(filter) = final_filter {
+                debug!(
+                    "using attribute filter {:?} for layer {:?}",
+                    &filter, &dataset_information.layer_name
+                );
+
+                // This is necessary because otherwise the GDAL postgres driver does not perform a filter-pushdown in case an explicit SQL query is given
+                format!("SELECT * FROM ({sql}) q WHERE {filter}")
```

Comment:
is this nested select expensive? Can you have a look what the query optimizer does her (EXPLAIN PLAN)?
Wouldn't it be possible to extend the `sql` query by extending its `where` clause?

Reply:
The plans seem to be equivalent. Of course, the inner query must not have something like a limit, then it would not be equivalent. However, this is also how GDAL handles spatial filters. It takes the SQL query as an inner query and adds the spatial filter to the outer query. Of course, it's also possible to extend the WHERE clause instead. The only advantage I see with this, though, is that one can then filter on columns which are not SELECTed in the SQL query through the attribute filter. However, you can already do this now in the explicit query you pass, so I don't really see this as an advantage and would personally prefer this easier approach without modifying the query string.